Add `core` and `alloc` over `std` Lints

bushrat011899 opened the pull request:
# Objective

- Fixes #6370
- Closes #6581

## Solution

- Added the following lints to the workspace:
  - `std_instead_of_core`
  - `std_instead_of_alloc`
  - `alloc_instead_of_core`
- Used `cargo +nightly fmt` with [item level use formatting](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#Item%5C%3A) to split all `use` statements into single items.
- Used `cargo clippy --workspace --all-targets --all-features --fix --allow-dirty` to _attempt_ to resolve the new linting issues, and intervened where the lint was unable to resolve the issue automatically (usually due to needing an `extern crate alloc;` statement in a crate root).
- Manually removed certain uses of `std` where negative feature gating prevented `--all-features` from finding the offending uses.
- Used `cargo +nightly fmt` with [crate level use formatting](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#Crate%5C%3A) to re-merge all `use` statements matching Bevy's previous styling.
- Manually fixed cases where the `fmt` tool could not re-merge `use` statements due to conditional compilation attributes.

## Testing

- Ran CI locally

## Migration Guide

The MSRV is now 1.81. Please update to this version or higher.

## Notes

- This is a _massive_ change to try and push through, which is why I've outlined the semi-automatic steps I used to create this PR, in case this fails and someone else tries again in the future.
- Making this change has no impact on user code, but does mean Bevy contributors will be warned to use `core` and `alloc` instead of `std` where possible.
- This lint is a critical first step towards investigating `no_std` options for Bevy.

alice-i-cecile commented:
I think we should do this. This is easy to enforce and for contributors to adapt to, and the auto-fix is particularly nice. Import consistency is a nice side benefit too.

I'm not convinced that Bevy will ever be able to be fully no-std, but doing this now will make supporting weird platforms (notably consoles) in the future much easier and give us a better sense of how far away we are, what changes we need upstream and what parts of the standard library the teams will need to reimplement for their platform.

BenjaminBrienen commented on crates/bevy_app/src/lib.rs:
```diff
@@ -7,6 +7,8 @@
 
 //! This crate is about everything concerning the highest-level, application layer of a Bevy app.
 
+extern crate alloc;
```
I don't think alloc is used here. Can this be removed?

bushrat011899 replied:
Unfortunately not, `bevy_app` uses `Cow` and `Rc` from `alloc`.

mockersf reviewed (dismissed):
I'm OK with doing this internally, but I don't think examples should be changed. They should still use `std`

bushrat011899 commented:
> I'm OK with doing this internally, but I don't think examples should be changed. They should still use `std`

I've switched the examples back to using `std` instead of `core` or `alloc`, and added the appropriate `#![expect(...)]` attributes to relevant examples so they still pass CI. I would _like_ to just have separate lints for the examples, but I couldn't see a quick way to implement it.

bushrat011899 commented:
To avoid changes to the examples, I've had to duplicate the workspace lints in the root `Cargo.toml`, as there is currently no way to override workspace lints within a particular member (see [rust-lang/cargo#13157](https://github.com/rust-lang/cargo/issues/13157) for details). Examples are no unchanged beyond their `use` statements being re-merged by `cargo +nightly fmt` as described in this PR description.

bushrat011899 commented on Cargo.toml:
```diff
@@ -10,7 +10,7 @@ keywords = ["game", "engine", "gamedev", "graphics", "bevy"]
 repository = "https://github.com/bevyengine/bevy"
 documentation = "https://docs.rs/bevy"
-rust-version = "1.79.0"
+rust-version = "1.81.0"
```
Required for `core::error::Error`

bushrat011899 commented on Cargo.toml:
```diff
@@ -45,15 +45,42 @@ ptr_as_ptr = "warn"
 ptr_cast_constness = "warn"
 ref_as_ptr = "warn"
 
+std_instead_of_core = "warn"
+std_instead_of_alloc = "warn"
+alloc_instead_of_core = "warn"
```
These lints direct Bevy contributors to use `core` where possible, then `alloc`, and then finally `std`. The status quo is to just import everything from `std`. Enabling these lints makes investigation into a `no_std` Bevy feasible. Additionally, items in `alloc` and `std` but not in `core` can indicate blocking or allocating types, which could impact performance.

bushrat011899 commented on Cargo.toml:
```diff
@@ -45,15 +45,42 @@ ptr_as_ptr = "warn"
 unsafe_op_in_unsafe_fn = "warn"
 unused_qualifications = "warn"
 
-[lints]
-workspace = true
```
We explicitly _don't_ want examples to be subject to the same linting requirements as the rest of Bevy, since it is encouraged that users rely on the Rust `std` library when possible. Unfortunately, there is no way to override just those lints, so a copy of the workspace lints is needed.

alice-i-cecile replied:
Please add a comment to this effect here and in the workspace lints to reduce the risk of desync.

bushrat011899 replied:
Done!

bushrat011899 commented on benches/benches/bevy_ecs/change_detection.rs:
```diff
@@ -86,7 +86,7 @@ fn generic_bench<P: Copy>(
 fn all_added_detection_generic<T: Component + Default>(group: &mut BenchGroup, entity_count: u32) {
     group.bench_function(
-        format!("{}_entities_{}", entity_count, std::any::type_name::<T>()),
+        format!("{}_entities_{}", entity_count, core::any::type_name::<T>()),
```
Most changes look like this, where some inline call to a `std` item needs to be replaced with the `core`/`alloc` alternative. As `std` re-exports `alloc` and `core`, this is a non-breaking change even in public facing signatures.

bushrat011899 commented on crates/bevy_a11y/src/lib.rs:
```diff
@@ -7,10 +7,10 @@
-    atomic::{AtomicBool, Ordering},
-    Arc,
-};
+extern crate alloc;
+
+use alloc::sync::Arc;
+use core::sync::atomic::{AtomicBool, Ordering};
```
In cases where the `alloc` crate is required, an `extern crate alloc;` must be placed in the `lib.rs`/`main.rs` file at the crate root. It may be possible to remove these if we instead created a `bevy_alloc` crate, and depend on it for the `alloc` types. This may be desirable anyway, as it would be a more appropriate home for things like `HashMap` and `EntityHashSet`.

james7132 replied:
I'd rather us just follow convention instead of establishing our own patterns. This seems like a rather non-intrusive requirement.

bushrat011899 replied:
I agree, and there's definitely a benefit to keeping it explicit and simple, especially considering this feature is aimed at the more low-level side of Rust programming.

bushrat011899 commented on crates/bevy_animation/src/lib.rs:
```diff
@@ -7,24 +7,27 @@
 use bevy_app::{App, Plugin, PostUpdate};
 use bevy_asset::{Asset, AssetApp, Assets, Handle};
 use bevy_core::Name;
 use bevy_ecs::{entity::MapEntities, prelude::*, reflect::ReflectMapEntities};
 use bevy_math::{FloatExt, FloatPow, Quat, Vec3};
-use bevy_reflect::std_traits::ReflectDefault;
-use bevy_reflect::Reflect;
+use bevy_reflect::{std_traits::ReflectDefault, Reflect};
```
Due to the use of `cargo +nightly fmt` to split and re-merge `use` statements, there are some additional changes beyond just the `std`, `alloc`, and `core` crates. It has been indicated in the `rustfmt.toml` file that _ideally_ we would already be using `imports_granularity = "Crate"` if it was a stable feature, so I don't consider this controversial.

bushrat011899 commented on crates/bevy_app/src/app.rs:
```diff
@@ -714,12 +711,12 @@ impl App {
     ///     })
     ///     // Registering an existing function with its type name
-    ///     .register_function_with_name(std::any::type_name_of_val(&mul), mul)
+    ///     .register_function_with_name(core::any::type_name_of_val(&mul), mul)
```
Documentation has also been updated where appropriate due to the linting requirements. This is visible to users of Bevy, but I think a reasonable change.

bushrat011899 commented on crates/bevy_app/src/schedule_runner.rs:
```diff
@@ -6,9 +6,11 @@ use crate::{
 #[cfg(target_arch = "wasm32")]
-use std::{cell::RefCell, rc::Rc};
-#[cfg(target_arch = "wasm32")]
-use wasm_bindgen::{prelude::*, JsCast};
+use {
+    alloc::rc::Rc,
+    core::cell::RefCell,
+    wasm_bindgen::{prelude::*, JsCast},
+};
```
Due to how `cargo +nightly fmt` split use statements, there were many instances of duplicated conditional compilation attributes. As a part of re-merging imports, I grouped these conditional imports together. I think this is easier to read, but may be contentious.

bushrat011899 commented on crates/bevy_asset/src/io/file/file_watcher.rs:
```diff
@@ -1,14 +1,14 @@
     new_debouncer,
     notify::{
-        self,
         event::{AccessKind, AccessMode, CreateKind, ModifyKind, RemoveKind, RenameMode},
-        RecommendedWatcher, RecursiveMode, Watcher,
+        RecommendedWatcher, RecursiveMode, Watcher, {self},
```
This change from `self` to `{self}` was also automatically done by the nightly formatter.

bushrat011899 commented on crates/bevy_derive/src/derefs.rs:
```diff
@@ -19,7 +19,7 @@ pub fn derive_deref(input: TokenStream) -> TokenStream {
 
     TokenStream::from(quote! {
-        impl #impl_generics ::std::ops::Deref for #ident #ty_generics #where_clause {
+        impl #impl_generics ::core::ops::Deref for #ident #ty_generics #where_clause {
```
Care needs to be taken with these macros, as we _can_ rely on the existence of `core`, but we _can't_ rely on the existence of `alloc`, as the callsite needs to already have `extern crate alloc;` added. Again, this would be addressed by providing a `bevy_alloc` crate, giving a concrete FQN that we can always rely on.

bushrat011899 commented on crates/bevy_ecs/examples/change_detection.rs:
```diff
@@ -6,6 +6,8 @@
 //! To demonstrate change detection, there are some console outputs based on changes in
 //! the `EntityCounter` resource and updated Age components
 
+#![expect(clippy::std_instead_of_core)]
```
These examples require this `expect` attribute because they are inside the `bevy_ecs` crate and thus subject to the workspace lints. It's not ideal, but there are only a couple of cases where this is required.

bushrat011899 commented on crates/bevy_ecs/src/label.rs:
```diff
@@ -1,10 +1,14 @@
     hash::{Hash, Hasher},
 };
 
+// Re-exported for use within `define_label!`
+#[doc(hidden)]
+pub use alloc::boxed::Box;
```
This is required for the `define_label!` macro. It's definitely not ideal, and I'd like to move it to a dedicated `bevy_alloc` crate instead, but it's a minimal issue IMO.

bushrat011899 commented on examples/2d/rotation.rs:
```diff
@@ -1,7 +1,6 @@
 
-use bevy::math::ops;
-use bevy::prelude::*;
+use bevy::{math::ops, prelude::*};
```
Examples have just had minor formatting changes made to their `use` statements based on `cargo +nightly fmt`. They are exempt from the lint requiring `core` and `alloc` where possible.

bushrat011899 reviewed (commented):
Below are some comments for reviewers which may help with parsing such a large set of changes. I believe this highlights all the noteworthy classes of changes that are made in this PR.

mockersf commented on examples/asset/multi_asset_sync.rs:
```diff
@@ -1,12 +1,10 @@
 use std::{
     f32::consts::PI,
     ops::Drop,
-    sync::{
-        atomic::{AtomicBool, AtomicU32, Ordering},
-        Arc,
-    },
+    sync::atomic::{AtomicBool, AtomicU32, Ordering},
```
why move `std::sync::Arc` outside of the other `std` imports?

bushrat011899 replied:
This was not a deliberate change, just a consequence of the `rustfmt` process used to address `clippy`. I'll update the PR to re-merge any items that got split out in examples.

mockersf commented on examples/window/screenshot.rs:
```diff
@@ -1,9 +1,10 @@
-use bevy_render::view::cursor::CursorIcon;
-use bevy_render::view::screenshot::{save_to_disk, Capturing, Screenshot};
+use bevy::{prelude::*, window::SystemCursorIcon};
+use bevy_render::view::{
+    cursor::CursorIcon,
+    screenshot::{save_to_disk, Capturing, Screenshot},
+};
```
probably for another PR: this example shouldn't use bevy_render directly. we used to have a CI that errored when that happens

bushrat011899 replied:
Addressed in #15333

mockersf commented on tests/window/change_window_mode.rs:
```diff
@@ -39,11 +39,11 @@ fn toggle_window_mode
         WindowMode::Windowed => {
             //it takes a while for the window to change from windowed to sizedfullscreen and back
-            std::thread::sleep(std::time::Duration::from_secs(4));
+            std::thread::sleep(core::time::Duration::from_secs(4));
```
```suggestion
            std::thread::sleep(std::time::Duration::from_secs(4));
```

mockersf commented on tests/window/change_window_mode.rs:
```diff
@@ -39,11 +39,11 @@ fn toggle_window_mode
         },
         _  => {
-            std::thread::sleep(std::time::Duration::from_secs(4));
+            std::thread::sleep(core::time::Duration::from_secs(4));
```
```suggestion
            std::thread::sleep(std::time::Duration::from_secs(4));
```

mockersf commented on examples/ui/text_debug.rs:
```diff
@@ -1,6 +1,7 @@
 //! Shows various text layout options.
 
-use std::{collections::VecDeque, time::Duration};
+use std::collections::VecDeque;
+use std::time::Duration;
```
```suggestion
use std::{collections::VecDeque, time::Duration};
```

mockersf commented on examples/shader/texture_binding_array.rs:
```diff
@@ -17,7 +16,8 @@ use bevy::{
     },
 };
-use std::{num::NonZero, process::exit};
+use std::num::NonZero;
+use std::process::exit;
```
```suggestion
use std::{num::NonZero, process::exit};
```

mockersf commented on examples/shader/gpu_readback.rs:
```diff
@@ -10,7 +10,9 @@
     render::{
-        render_graph::{self, RenderGraph, RenderLabel},
+        render_graph::{
+            RenderGraph, RenderLabel, {self},
```
same here for `{self}`

bushrat011899 commented on examples/shader/specialized_mesh_pipeline.rs:
```diff
@@ -29,7 +29,9 @@ use bevy::{
         texture::BevyDefault as _,
-        view::{self, ExtractedView, ViewTarget, VisibilitySystems, VisibleEntities},
+        view::{
+            ExtractedView, ViewTarget, VisibilitySystems, VisibleEntities, {self},
```
No this is not needed. This was done by the `cargo +nightly fmt` formatter. As it's ugly, I'll undo those changes and use the previous style.

james7132 reviewed (commented):
Only have looked through half the the files so far, but so far LGTM.

mockersf commented on examples/games/contributors.rs:
```diff
@@ -2,9 +2,10 @@
 use bevy::{math::bounding::Aabb2d, prelude::*, utils::HashMap};
 use rand::{prelude::SliceRandom, Rng};
+use std::hash::{Hash, Hasher};
 use std::{
     env::VarError,
-    hash::{DefaultHasher, Hash, Hasher},
+    hash::DefaultHasher,
```
another one to revert

mockersf commented on crates/bevy_utils/src/short_names.rs:
```diff
@@ -1,7 +1,7 @@
 ///
 /// The short name of a type is its full name as returned by
-/// [`std::any::type_name`], but with the prefix of all paths removed. For
+/// [`core::any::type_name`], but with the prefix of all paths removed. For
```
can we keep references to `std` in doc comments or does that trigger the lint?

bushrat011899 replied:
Unfortunately yes, the lint does _not_ catch using `std` in documentation links or examples. As such I will revert changes to documentation since we can't enforce the use anyway.

However, `no_std` crates like `bevy_utils` and `bevy_ptr` _need_ to link to `alloc` and `core` since `std` is unavailable.

cart commented on Cargo.toml:
```diff
@@ -45,15 +45,53 @@ ptr_as_ptr = "warn"
 ptr_cast_constness = "warn"
 ref_as_ptr = "warn"
 
+std_instead_of_core = "warn"
```
The downsides I can see are:

1. This is a large / sweeping change that permanently changes how we all write Bevy code.
2. This introduces _slightly_ more boilerplate (we need `extern crate alloc` in most crates now).
3. This takes us further from the "internal Bevy code looks like user-facing Bevy code" principle, because the majority of users / plugins can (and should) continue to use `std`.

As such, before making such a change, it would be good to qualify _exactly_ what it wins us. For example, getting closer to some _potential_ `no_std` Bevy future doesn't do us much if there are true blockers that prevent us from finishing that journey. Same goes for console support (which is currently phrased like "this _might_ help").

Before merging, I'd want satisfying answers to:

1. With this change, what are the remaining blockers to "no_std Bevy"
2. With _just_ this change, what scenarios are we unlocking?
3. For (2), can we unlock those scenarios with targeted ports of specific crates? (ex: doing this for bevy_ecs)

I also question the viability of this effort for things like consoles when the majority of plugins and user code will still target `std`. Seems like an `std` console port (even if it starts as a partial / shimmed port) would be preferable from an ecosystem perspective.

bushrat011899 replied:
Thanks for taking the time to look at this! As an immediate response, here's what I personally think in regards to those listed downsides:

> 1. This is a large / sweeping change that permanently changes how we all write Bevy code.

Agreed, this will affect _all_ contributors for Bevy going forward. However, I believe this is a relatively minor impact, as `clippy` can suggest the exact changes required to pass these lints. In effect, contributors can continue to write `std`-reliant code, but on completion they'll fix this lint the same way we already do for things like `fmt` and `taplo`.

> 2. This introduces slightly more boilerplate (we need `extern crate alloc` in most crates now).

This is true, effectively every crate root (`lib.rs`) will include `extern crate alloc`. However, it is only a single line of code per crate, so I think, while annoying, its real impact is negligible. Especially considering this PR adds the lines already, so contributors will only encounter this boilerplate when creating a new Bevy crate.

> 3. This takes us further from the "internal Bevy code looks like user-facing Bevy code" principle, because the majority of users / plugins can (and should) continue to use `std`.

This is true and I don't have a meaningful rebuttal; this will add one layer of indirection between user and contributor code. What I would say is that we already have that for things like `bevy_reflect` being feature-gated. The majority of users write code for Bevy assuming `bevy_reflect` is available, but contributors _can't_ make that same assumption. 

> As such, before making such a change, it would be good to qualify exactly what it wins us. For example, getting closer to some potential `no_std` Bevy future doesn't do us much if there are true blockers that prevent us from finishing that journey. Same goes for console support (which is currently phrased like "this _might_ help").

I personally don't have access to any modern console SDKs so _I_ can't comment on the necessity for a `no_std` Bevy for platforms like the Nintendo Switch, Xbox Series, PS5, etc. However, I _can_ say that for the homebrew and retro community, `no_std` is an [absolute must](https://github.com/bevyengine/bevy/discussions/10680). For some context on my personal motivations here, I'm wanting to use the [`agb` crate](https://crates.io/crates/agb) with Bevy to create some games for the GameBoy Advance. To do this, I need the following crates to be `no_std` at a minimum:

- [X] `bevy_ptr`
- [X] `bevy_utils`
- [X] `bevy_tasks`
- [ ] `bevy_ecs`
- [ ] `bevy_app`

Out of the above 5 crates, `bevy_ptr` and `bevy_utils` are _already_ `no_std`, and I have a [draft `no_std` for `bevy_tasks`](https://github.com/bevyengine/bevy/pull/15362) already published. `bevy_app` is trivial to make `no_std`, as it only uses the standard library for `thiserror`, `Mutex`, and `panic` capturing, all of which have suitable `no_std` replacements (or just not-applicable in the case of `panic` capturing). This leaves `bevy_ecs` as the only core crate remaining, and I'm already about half-way through a port of that too, with the biggest issues being non-`Send` resources (requires access to `std::thread`) and the schedule graph using a single type and algorithm from `petgraph` (`GraphMap`).

Once the above changes are made, I will be able to use Bevy on _any_ platform, including the GameBoy Advance. What's more, I'd be able to use the high level abstractions like systems, queries, and the `App` and plugin infrastructure too, making Bevy the single best engine for retro-consoles.

With the above context, here are my answers to your three questions:

> 1. With this change, what are the remaining blockers to "no_std Bevy"

1. `thiserror`
2. `petgraph`'s `GraphMap` data structure
3. non-`Send` resources
4. Refactoring `bevy_ecs` and `bevy_app` to include a `std` feature

Obviously this excludes other aspects of Bevy, (`bevy_time`, etc.) which I _want_ to be `no_std`, but everything else is optional, `bevy_app` and `bevy_ecs` are the minimum to provide an end-user appropriate `no_std` experience.

> 2. With just this change, what scenarios are we unlocking?

To make a crate like `bevy_ecs` `no_std`, the first step will be this very refactor, and then it'll be _meaningful_ work (e.g., `petgraph`, etc.). The problem is that any PR which attempts to make `bevy_ecs` `no_std` will be made vastly more controversial by mixing this PR's work _and_ their actual changes to Bevy's code, making review substantially harder and more likely to fail. By applying this lint in its own PR and across the workspace, we remove that controversy and noise from future discussions around `no_std` Bevy, and we keep all of Bevy consistent (_"Oh you can't write that code here, we have the `std` lint on"_).

> 3. For (2), can we unlock those scenarios with targeted ports of specific crates? (ex: doing this for `bevy_ecs`)

Yes, but I believe this introduces the very inconsistency in Bevy that we'd like to avoid. These lints make no restrictions around what code we write, or even how they're written, only the names of `core` and `alloc` types. These lints provide a consistent experience for all Bevy contributions, regardless of whether you're in a `no_std` crate or not. Since Bevy already has two `no_std` crates, these lints will make working on those crates consistent with the rest of the library.

> I also question the viability of this effort for things like consoles when the majority of plugins and user code will still target std.

I see this as a chicken-and-egg problem; Bevy's plugins are `std` because Bevy is `std`. You actually _can't_ write a `no_std` Bevy plugin right now because the required traits are "stuck" inside an `std`-only crate. I can personally attest to my intention to make a `bevy_agb` plugin to allow working on the GameBoy Advance, and I'm certain that other community members would have similar excitement around a new area of Bevy to explore. Bare-metal games on the Raspberry Pi, Playdate support, embedded projects using `bevy_ecs` as a database, etc. are projects I have seen discussed in the Bevy Discord as fantastical, but I genuinely believe are achingly close to realisation.

> Seems like an std console port (even if it starts as a partial / shimmed port) would be preferable from an ecosystem perspective.

For the modern consoles, definitely, but this would not be viable for anything retro. Additionally, with Bevy relying on the whole `std` (right now), knowing how much of the `std` library you need to port is difficult. If Bevy's reliance on the standard library was reduced to what it _needs_, then that work also becomes more actionable.

I apologise for the long response, especially considering just how large this PR itself is, and I thank you for taking the time to read it!

alice-i-cecile replied:
> thiserror

I'd like to keep this if possible, I think with `core::Error` stabilizing we can actually fix this upstream.

> petgraph's GraphMap data structure

Not long for this world frankly, the ECS folks kinda hate petgraph

> non-Send resources

Feature-flagging this is fine

> Refactoring bevy_ecs and bevy_app to include a std feature

Very doable.

bushrat011899 replied:
Ok, to provide some hard evidence for how close `no_std` Bevy is, I have made some rough ports of `bevy_ecs` and `bevy_app` on [this branch](https://github.com/bushrat011899/bevy/tree/NoSTDBevyECS). All that really had to change was adding an `Instant` type appropriate for `no_std` environments, and porting `petgraph`'s required parts to a custom solution. That in of itself was actually just an almost straight copy-paste of `petgraph`'s implementation (simplified), since it was already _almost_ `no_std`, just needed to disable a couple features we weren't using.

These changes allow this application to compile on `x86_64-unknown-none`, bare-metal x86:

<details>
  <summary>`Cargo.toml`</summary>
  
```toml
[package]
name = "bevy_no_std_test"
version = "0.1.0"
edition = "2021"

[dependencies]
bevy_app = { path = "../bevy_app", version = "0.15.0-dev", default-features = false }
bevy_ecs = { path = "../bevy_ecs", version = "0.15.0-dev", default-features = false }

[profile.dev]
opt-level = 3
debug = true
panic = "abort"
```
  
</details>


<details>
  <summary>`main.rs`</summary>
  
```rust
//! Demonstrates a [`no_std`] Bevy application.

#![allow(unsafe_code)]
#![no_std]
#![no_main]

// Since we don't have access to `std`, and the `core` library can't assume there's
// a global allocator, we have to explicitly include it using `extern crate`.
extern crate alloc;

// Explicitly importing bevy sub-crates because I haven't threaded the `std` feature
// through bevy_internal and then bevy proper.
use bevy_app::prelude::*;
use bevy_ecs::prelude::*;

#[derive(Resource, Default)]
struct MyResource {
    hello_world: bool,
}

#[derive(Component)]
struct Player;

fn main() {
    App::empty()
        .init_resource::<MyResource>()
        .add_systems(Update, |my_resource: Res<MyResource>| {
            assert_eq!(my_resource.hello_world, false);
        })
        .add_systems(Startup, |mut commands: Commands| {
            commands.spawn(Player);
        })
        .add_systems(Update, |query: Query<&Player>| {
            assert_eq!(query.iter().count(), 1);
        })
        .run();
}

/// You cannot rely on [`main`] being called for you in a `no_std` environment.
/// The exact name of your start function will depend on the platform you're developing for.
/// It's convention to name this function `_start`, and we use `#[no_mangle]` to ensure
/// the Rust compiler doesn't change this name.
#[no_mangle]
pub extern "C" fn _start() -> ! {
    // Our start function is just going to call main, and if main exits we'll
    // just loop forever. Quitting in a `no_std` context isn't defined, since
    // you might be running on bare-metal hardware without power management!
    main();
    loop {}
}

/// Bevy requires access to a global allocator, another item which is platform specific
/// and so must be setup manually.
#[global_allocator]
static ALLOCATOR: allocator::Dummy = allocator::Dummy;

/// Obviously _your_ code will never panic. But, Rust still requires setting up
/// a panic handler _just in case_.
#[panic_handler]
fn panic(_info: &core::panic::PanicInfo) -> ! {
    loop {}
}

/// Custom allocator for our no_std environment
mod allocator {
    use alloc::alloc::{GlobalAlloc, Layout};
    use core::ptr::null_mut;

    /// This is a demonstration allocator and is _intentionally_ broken.
    pub struct Dummy;

    unsafe impl GlobalAlloc for Dummy {
        unsafe fn alloc(&self, _layout: Layout) -> *mut u8 {
            null_mut()
        }

        unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {
            panic!("dealloc should be never called")
        }
    }
}
```
  
</details>

After applying the lints in this PR, it's about another 3000 diff-lines to make `bevy` usable in `no_std` applications with all the high-level functionality people rely on Bevy for.

BenjaminBrienen replied:
I think this is a big win for bevy as the *core* :wink: of the ecosystem for a very small price.

cart replied:
Ok I'm convinced. Lets do this!

bushrat011899 replied:
Brilliant! I'm going to resolve the merge conflicts now so hopefully this could be merged soon!

alice-i-cecile commented:
@bushrat011899 Ping me when merge conflicts are resolved and I'll get this in ASAP.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1698 if you'd like to help out.
